hack: add apidiff.sh

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Making changes in staging is always a risk regarding incompatible API changes. apidiff helps detect those. The next step is adding a Prow job which invokes it.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

apidiff can be invoked for a single internal package or iterate over everything, including staging. The base to compare against can be specified via -r (similar to verify-golangci-lint.sh), with the default the base revision on master (similar to -a in verify-golangci-lint.sh).

Example:

```patch
diff --git a/staging/src/k8s.io/client-go/rest/warnings.go b/staging/src/k8s.io/client-go/rest/warnings.go
index ad493659f22..47cdf9854a1 100644
--- a/staging/src/k8s.io/client-go/rest/warnings.go
+++ b/staging/src/k8s.io/client-go/rest/warnings.go
@@ -31,6 +31,9 @@ import (
 type WarningHandler interface {
        // HandleWarningHeader is called with the warn code, agent, and text when a warning header is countered.
        HandleWarningHeader(code int, agent string, text string)
+
+       // Foo is not compatible
+       Foo()
 }
 
 var (
diff --git a/test/e2e/framework/providers/gce/ingress.go b/test/e2e/framework/providers/gce/ingress.go
index 1e4a8e649bd..28ea5991f8f 100644
--- a/test/e2e/framework/providers/gce/ingress.go
+++ b/test/e2e/framework/providers/gce/ingress.go
@@ -176,14 +176,6 @@ func (cont *IngressController) deleteForwardingRule(del bool) string {
        return msg
 }
 
-// GetGlobalAddress returns the global address by name.
-func (cont *IngressController) GetGlobalAddress(ipName string) *compute.Address {
-       gceCloud := cont.Cloud.Provider.(*Provider).gceCloud
-       ip, err := gceCloud.GetGlobalAddress(ipName)
-       framework.ExpectNoError(err)
-       return ip
-}
-
 func (cont *IngressController) deleteAddresses(del bool) string {
        msg := ""
        ipList := []compute.Address{}
@@ -439,6 +431,8 @@ func (cont *IngressController) ListInstanceGroups() []*compute.InstanceGroup {
        return igList
 }
 
+func (cont *IngressController) Foo() {}
+
 func (cont *IngressController) deleteInstanceGroup(del bool) (msg string) {
        gceCloud := cont.Cloud.Provider.(*Provider).gceCloud
        // TODO: E2E cloudprovider has only 1 zone, but the cluster can have many.
```

Output:
```
$ hack/apidiff.sh 
go version go1.21.6 linux/amd64
installing apidiff into /nvme/gopath/src/k8s.io/kubernetes/_output/local/bin
Preparing worktree (detached HEAD 6892e38362a)
Updating files: 100% (24026/24026), done.
HEAD is now at 6892e38362a Merge pull request #122963 from SataQiu/fix-20240125

# k/k
Incompatible changes:
- ./test/e2e/framework/providers/gce.(*IngressController).GetGlobalAddress: removed
Compatible changes:
- ./test/e2e/framework/providers/gce.(*IngressController).Foo: added

# staging/src/k8s.io/client-go
Incompatible changes:
- ./rest.WarningHandler.Foo: added
```


#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/assign @dims

pohly committed:
hack: add apidiff.sh

apidiff can be invoked for a single internal package or iterate over
everything, including staging. The base to compare against can be specified via
-r (similar to verify-golangci-lint.sh), with the default the base revision on
master (similar to -a in verify-golangci-lint.sh).

dims commented:
cc @BenTheElder @liggitt

pohly commented:
Currently the script can be run manually. The next step will be an optional, non-blocking `pull-kubernetes-apidiff` Prow job which invokes the script. We could also add some "run if changed" rules, but I am not sure what those should be.

pohly commented:
/sig testing

liggitt commented:
Is this going to be useful? I suspect a majority of PRs will change an exported go function signature under some staging package. Nobody will pay attention to an optional presubmit that is ~usually red.

Are there key packages / interfaces that it would make more sense to pay closer attention to? Things like client-go client / informer packages, core apimachinery packages, etc?

pohly commented:
> Is this going to be useful?

I am currently trying to make client-go support contextual logging. This is only possible via API changes (or rather, extensions). In that context I would find it useful to run apidiff to show to myself and others what is and what isn't a breaking API change. I usually spot them during code review, but others don't (adding methods to interfaces, changing a function signature by adding optional parameters are both recent examples).

> I suspect a majority of PRs will change an exported go function signature under some staging package. Nobody will pay attention to an optional presubmit that is ~usually red.

But perhaps they should? I also question the "majority of PRs will change" and "usually red" - I think we are more careful about not breaking consumers of our staging repos unless there's a good reason and it cannot be avoided.

> Are there key packages / interfaces that it would make more sense to pay closer attention to? Things like client-go client / informer packages, core apimachinery packages, etc?

Those would be good candidates for "run if changed". But we can leave that out and gather some experience with the script on some PRs before taking that as a next step. I'm also fine with keeping it something that I trigger manually.

pohly commented:
There obviously wasn't much interest in having this in-tree. Let's close.

/close

pohly commented:
/reopen
/assign @aojea

pohly commented on hack/apidiff.sh:
```diff
@@ -0,0 +1,152 @@
+    compare "${targets[*]}" "${KUBE_TEMP}/before/all.out" "${KUBE_TEMP}/after/all.out"
+else
+    compare "k/k" "${KUBE_TEMP}/before/all.out" "${KUBE_TEMP}/after/all.out"
+    for d in staging/src/k8s.io/*; do
```
This is still from the pre-workspace era. I'll simplify it.

/hold

pohly commented on hack/tools/go.mod:
```diff
@@ -10,6 +10,7 @@ require (
 	github.com/jcchavezs/porto v0.6.0
 	go.uber.org/automaxprocs v1.5.2
 	go.uber.org/mock v0.4.0
+	golang.org/x/exp v0.0.0-20240416160154-fe59bbe5cc7f
```
@liggitt: You were concerned about adding a dependency here.

Perhaps it would be better to always install the latest apidiff? Then we don't need to bother with updating this dependency.

There's no guarantee that the latest release will work, but that IMHO is okay for this script (optional, not required to work).

liggitt replied:
we don't even have to install latest that way, right? we could bake a specific version into the install command. I think we did that with zeitgeist as well for the same reason in https://github.com/kubernetes/kubernetes/pull/118076/files

pohly replied:
apidiff doesn't have tagged versions.

liggitt replied:
can't we use a pseudoversion to install?

pohly replied:
I suppose. But do we want to to use a fixed revision instead of always using "the latest"?

It will be another, non-obvious thing that needs to be bumped occasionally, probably when someone notices a problem. This is bound to happen at some point, whereas always using the latest might never cause problems. In klog, we've always used the latest without problems in the last five years.

OTOH, there is a certain (low?) risk that a future revision does something nasty in our build environment.

:shrug:

pohly replied:
> ... in our build environment.

Or on a developer's machine.

pohly replied:
I've pushed a version which uses `go install @latest`

liggitt replied:
if this is advisory and non-blocking and won't be an issue in release branches, I can live with `@latest`. If this was blocking and would get branched into release branches and live there, I'd want to pin.

pohly replied:
Check:
- [x] not blocking
- [x] not branched

Can you LGTM or do you want to leave that to someone else?

/hold

To prevent accidental merge (is self-approved).

pohly commented:
@aojea: as you were interested in using this, perhaps you can try it out?

I just pushed an update that works with workspace mode.

A simple test is:
```patch
diff --git a/staging/src/k8s.io/client-go/tools/metrics/metrics.go b/staging/src/k8s.io/client-go/tools/metrics/metrics.go
index 99d3d8e239c..31fd3c20a6d 100644
--- a/staging/src/k8s.io/client-go/tools/metrics/metrics.go
+++ b/staging/src/k8s.io/client-go/tools/metrics/metrics.go
@@ -124,6 +124,8 @@ type RegisterOpts struct {
        RequestRetry          RetryMetric
        TransportCacheEntries TransportCacheMetric
        TransportCreateCalls  TransportCreateCallsMetric
+
+       NonCmp func()
 }
 
 // Register registers metrics for the rest client to use. This can
```

```console
$ ./hack/apidiff.sh ./staging/src/k8s.io/client-go/
installing apidiff into /tmp/tmp.9ZoXbo5YeG
Preparing worktree (detached HEAD f39ece24b27)
Updating files: 100% (24395/24395), done.
HEAD is now at f39ece24b27 Merge pull request #124393 from carlory/kubeadm-3045-1

## ./staging/src/k8s.io/client-go/
Incompatible changes:
- ./tools/metrics.RegisterOpts: old is comparable, new is not
Compatible changes:
- ./tools/metrics.RegisterOpts.NonCmp: added

```

pohly commented:
/retest

pohly committed:
hack: use "latest" apidiff

Pinning apidiff to a specific version shouldn't be necessary because, if past
experience in klog holds true, the latest version just works. This way we don't
have to remember to bump up the revision.

The downside of using "latest" is that a compromise of that version would give
an attacker the ability to run code in the CI and on developer machines.
